### ISSUE TITLE
Fix algorithm inconsistencies

### ISSUE THREAD

**Issue by ddorwin** [w3c/encrypted-media#181]
This issue tracks a set of issues that affect spec and algorithmic correctness and readability. (Many in this initial list were discussed and/or identified in #171 and related PRs but are independent of or should be resolved before #171.) Please feel free to propose related additional candidates in comments. To facilitate more effective reviews, please submit focused PRs for one or more related issue and/or algorithm.

1. #153
1. Clearly document which entity (user agent or CDM) executes each algorithm to reduce confusion and the potential for future issues. (This should be implemented along with #153.)
 * It may also be helpful to specify which entity calls an algorithm where appropriate.
 * For example, "This algorithm is [called by the CDM and] executed in the user agent."
1. Ensure the spec's use of algorithms is consistent with the above changes.
1. Make the Close All Sessions algorithm consistent with the `close()` algorithm (see https://github.com/w3c/encrypted-media/issues/171#issuecomment-217001207).
 * The Close All Sessions algorithm is currently called by both the user agent and CDM.
 * Other than the one case mentioned in the above comment, the Session Close algorithm is only called by the CDM.
1. Be consistent in how the user agent uses the CDM to perform operations (https://github.com/w3c/encrypted-media/issues/171#issuecomment-217001207).
1. Remove the redundant call to `update-key-statuses-algorithm` (https://github.com/w3c/encrypted-media/pull/179#discussion-diff-62074223).
1. Ensure the meaning of "the Session Close algorithm has been run on this object" is unambiguous.
 * Since this can be triggered by various paths and the algorithm itself may (TBD whether it does after fixing the above issues) involve asynchronous operations, it could be unambiguous as to at which point algorithm steps containing the condition above should be triggered.

<br/>Related but not specifically covered by this issue since it involves modifying behavior:
- Address the redundancy between `remove()` and `close()` described in #180.

**Comment by mwatson2** [comment]
I think we need to be careful with language such as "called by the CDM" or "executed by the CDM", since those phrases include execution model assumptions which may not hold in all implementations.

Clearly, there is functionality that is the responsibility of the CDM (including, at least, all the stuff we do not specify such as message formats). For any functionality that we _do_ specify, what does it really mean for us to assign that to the CDM vs the User Agent ? In any given implementation there are likely at least two software components, one of which contains the unspecified CDM functionality. However, the share of specified functionality in the same software component as the unspecified stuff may differ from implementation to implementation.

I think the most we can do is distinguish between steps that are triggered entirely by fully specified functionality (for example, method calls on the API) and steps that may be triggered by functionality that is not specified (sessions being closed for some CDM-specific reason).

I don't think we should speak about which component "executes" the steps or "calls" the algorithm.

I think all the issues raised above could be re-phrased in this form.

**Comment by ddorwin** [comment]
The intent is not to specify an execution model or communication mechanism. However, I think the spec should continue to have a clear conceptual model, regardless of how it might be implemented [1].

The spec is mostly consistent in defining that the user agent modifies the JS objects, firing events, etc. while the CDM manages keys, etc. There are a few cases where the algorithms have the CDM directly executing an algorithm that falls in the former category. I think that is confusing. Language similar to [this statement in your PR](https://github.com/w3c/encrypted-media/pull/179/commits/fc4d686c4c860741ad2d8aae78b6edf8144fa0df#diff-f72607e47a6f74e53dc90eab8ee094e2R2129) would help maintain consistency and improve readability.


I agree that distinguishing between _algorithms_ that are triggered entirely by fully specified functionality makes sense and is similar to the last point in #153. One possible option would be to have most algorithms only be called explicitly then have a few algorithms/statements that define calls to those algorithms in cases that are not specified by algorithms.

[1] The two components don't even need to be separate objects in the implementation. For example,  Chrome's Clear Key implementation is implemented directly in the user agent. While it is implemented behind a common CDM interface, the process model is different from other CDMs.

**Comment by mwatson2** [comment]
Ok, so I think we are on the same page. I do think that addressing the points here will also address #171, but we will see when we've done it.

I think the "Use the CDM to XYZ" language is good enough for calling out when unspecified functionality is invoked and at these points we can also place requirements on the CDM ("XYZ must result in PQR") for example "Closing the session MUST result in destruction of keys or licenses which have not been explicitly stored".

Would it make sense to have a section "CDM triggered asynchronous events" that lists the cases where CDM functionality can trigger one of our algorithms ? I think the cases are:
- spontaneous session close by CDM (new Session Closed algorithm as in #172 )
- Queue a Message Event (existing algorithm)
- Update Key Statuses (existing algorithm)
- Update Expiration (existing algorithm)

I happy to make a PR along these lines.

**Comment by ddorwin** [comment]
Destroying the keys is certainly an assumption, though really part of the Key System's contract, whenever the CDM closes a session other than `"persistent-license"`. Where the spec explicitly requests the CDM to close the session, we can be specific about this. I'm not sure how best to handle this for the "asynchronous" case. That might depend on how we decide to document such behaviors.

A new section (a peer to Algorithms?) for the asynchronous triggers seems sensible. (I'd avoid overloading the term "events.") We can always move it around, but it seems clear that we need some explicit language outside the existing algorithms. Such a change might be smaller if we address some of the other item first. (For example, I think this PR could grow too big if we try to refactor the session closing algorithms at the same time.)

I think we should be clear, though, that such triggers are the exception and most such calls are already normatively defined. This section only addresses things that are truly unrelated to actions by the application. It seems that the last three bullets above would only be called in such a way in response to policies in the license. For example, the renewal request period is reached, there is a change in output protection status, and the license expires, respectively. As counterexamples, we explicitly handle updating the key statuses or closing all sessions in the Attempt to Decrypt algorithm.

**Comment by steelejoe** [comment]
Any type of CDM denial-of-service seems like it could be expressed as a "CDM triggered asynchronous event". For example an out-of-process CDM crashing. Another example might be if the user changed the global policy settings around CDM usage while the CDM was in active use. I guess I am arguing for the 4th category @mwatson2 listed above. The overall approach seems sound.

**Comment by mwatson2** [comment]
@ddorwin 

> A new section (a peer to Algorithms?) for the asynchronous triggers seems sensible. (I'd avoid overloading the term "events.")

Ok.

> Such a change might be smaller if we address some of the other item first. (For example, I think this PR could grow too big if we try to refactor the session closing algorithms at the same time.)

Making it clear which algorithms are triggered by the specification and which by unspecified CDM behavior at least requires splitting of the Session Close algorithm, since session close can happen both ways and the procedures are different. But this is a very simple split into _Close Session_ and _Session Closed_ similar to #172.

We _could_ specify that _Queue a Message_, _Update Key Status_ and _Update Expiration_ are only ever triggered by the specification (including in the _Attempt to Decrypt_ algorithm). The practical effect is that things like license expiration and output protection changes are not indicated to the application unless we are actively decrypting data.

Alternatively, if these algorithms can be triggered at any time, these notifications could come immediately.

**Comment by ddorwin** [comment]
@mwatson2:

Splitting or otherwise refactoring the Session Close algorithm is fine. I was just saying we should make such a change in a separate small PR.

I was not suggesting we not allow the algorithms to be called immediately, just that _most_ calls are already normatively specified and implementations should prefer those normative calls when appropriate. There are unlikely to be observable differences between the normative steps in the Attempt to Decrypt and truly asynchronous calls except when playback is paused or there is a long series of unencrypted blocks. One could probably make the case that if we added a call to Update Expiration from Attempt to Decrypt that there is no unspecified reason that it would be called.

Another approach would be to define some sort of busy-loop algorithm that represents the asynchronous calls and state that it is always running but only as a model and not actually intended to be implemented that way. Then we could have a simple algorithm with normative steps that call the individual algorithms when necessary. For example, "1. If the status of any key has changed, run the Update Key Status algorithm." This is essentially a way of replacing the prose in the proposed  "CDM triggered asynchronous events" section with an algorithm, which seems desirable.

**Comment by mwatson2** [comment]
@ddorwin I like the idea of a busy-loop algorithm to specify the asynchronous triggers.

Still, I'm not sure this can be done without splitting Session Close in the same PR. The busy-loop will handle the case of "If _cdm_ has closed a session, then X" where X is not all of the current Session Close algorithm. I could duplicate the steps and have a separate PR to de-dup and re-factor, I guess.

**Comment by ddorwin** [comment]
@mwatson2, I'm not trying to make busy work or enforce some hard rule. It just seemed that we could, for example, define Close Session and Session Closed and reference them from the appropriate points in one PR then implement the busy-loop in a separate PR. (Looking at the spec, it seems Close All Sessions and `close()` would reference the former, and everything else would reference the latter.) This way, we can focus on the definition of those separate from, for example, the design of the busy-loop.

IMO, it is difficult to follow multiple rounds of feedback in various areas in GitHub; I'm hoping that such focused smaller changes will make this easier and help us reach a final state on each change quicker.

**Comment by mwatson2** [comment]
I agree it would be good to break up the changes and doing it that way around could work. I'm not sure that both PRs can exist simultaneously, though, unless the second one is a PR against the branch for the first one. Is that what you mean ?

**Comment by ddorwin** [comment]
That is one option. Or we can serialize them. Turnaround should be quicker for small PRs. :)

It's probably also possible to implement the busy-loop algorithm with stub references, review that in parallel, and fix it up once the first PR lands.

**Comment by mwatson2** [comment]
I created PR #184 to refactor Session Close.

I have also created [this PR ](https://github.com/mwatson2/encrypted-media/pull/1 )against the branch for the above PR (i.e. in my fork) to create a new "Monitor for CDM Changes" algorithm.

These PRs address items 1, 2, 3 and 6 in the OP. (2 is addressed because all algorithms are now executed by the same entity).

My proposal for the other items is as follows:
4. (Close All Session) This is only used for the "CDM becomes unavailable" case. It is triggered explicitly by the _Attempt To Decrypt_ and implicitly in the `MediaKeys` section "if the CDM becomes unavailable". I suggest we remove that implicit trigger and either:
(a) Do nothing (CDM unavailability will be detected only at decryption time), or
(b) Add a further step to the new _Monitor for CDM Changes_ that closes the session if the CDM becomes unavailable.

My preference is for (b).

5. Requires a separate PR to consistently use the "Use _cdm_ to ..." language and to be consistent in what the CDM is asked to do when a session is closed.

7. I think this is fine as it is ("If the session closed algorithm has been run...") but it could be changed to "If the `closed` attribute of _session_ has been resolved."

**Comment by mwatson2** [comment]
I believe the [first PR](https://github.com/w3c/encrypted-media/pull/184) for this issue is ready to merge.

Please review the [second PR](https://github.com/mwatson2/encrypted-media/pull/1) for this issue, which codifies the execution of algorithms triggered by the CDM. This has been updated to address item (4).

I created a [third PR](https://github.com/mwatson2/encrypted-media/pull/2) for this issue to address item (5).

The second and third PRs are written against the first (i.e. against that branch in my fork). I believe all the items in the OP are addressed by these three PRs.

Comments please.

**Comment by mwatson2** [comment]
Since it is relatively small and further discussion depends on it, I will merge the [first PR](https://github.com/w3c/encrypted-media/pull/184) if there are no further comments by end-of-day tomorrow.

**Comment by mwatson2** [comment]
I have rebased the second PR for this issue on the gh-pages branch. Please see this [new PR](https://github.com/w3c/encrypted-media/pull/205) instead.

**Comment by mwatson2** [comment]
The third PR for this issue addressed item (5) which is now addressed by the PR for issue #14. So, there is only one outstanding PR for this issue, referenced in the comment above.

**Comment by mwatson2** [comment]
Here is an [htmldiff](https://services.w3.org/htmldiff?doc1=http%3A%2F%2Frawgit.com%2Fmwatson2%2Fencrypted-media%2Ff8636f760d1d230403990bfca8ec59967729b30e%2Fencrypted-media-respec.html&doc2=http%3A%2F%2Frawgit.com%2Fmwatson2%2Fencrypted-media%2Fff15980d84cf7f5ecbde96a9a881840070779b2e%2Fencrypted-media-respec.html) for the remaining proposed changes for this issue.

**Comment by mwatson2** [comment]
@ddorwin @mwatson2 @jdsmith3000 Discussed at editors' meeting. Proposal is to specify that this algorithm is run "in parallel" to the event loop and that for each algorithm that is triggers we should "queue a task" to run that algorithm.

Need to clarify that this algorithm should be serialized with the "in parallel" blocks in the methods.

**Comment by mwatson2** [comment]
I believe all the PRs for this issue are now merged.

**Comment by ddorwin** [comment]
There are at least some minor things still to address. I created PR #217 for one category. I'll review this again for the rest.

**Comment by mwatson2** [comment]
Is there anything left on this one now #217 is merged ?

**Comment by ddorwin** [comment]
I'm going to review and verify after wrapping up the remaining V1 issues.

**Comment by ddorwin** [comment]
All of the issues in the original comment appear to have been addressed except for #153, which I have reopened.
